---
title: "Assignment 2 - BCB420"
subtitle: "Differential Gene Expression and Preliminary ORA"
author: "Alcides Albuquerque"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: TRUE
    toc_depth: 2
    df_print: paged
bibliography: A2.bib
csl: apa.csl
---

Quick links (more details at introduction):

Assignment used the base docker image [acessible here] (https://hub.docker.com/repository/docker/risserlin/bcb420-base-image).

Chosen study: [Serotonin-induced hyperactivity in SSRI-resistant major depressive disorder patient-derived neurons] (https://pubmed.ncbi.nlm.nih.gov/30700803/)[@vadodaria2019serotonin].

GEO data set: [GEO accession GSE125664; Vadoraria et al., 2019] (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125664)[@vadorariaGEO]

## 1. Introduction: A1 summary

## 1.1 Selection of an expression data set and software tools

Previously, in A1, we have found an appropriate study (in terms of RNA expression analysis and clinical relevance) by searching RNAseq data set available at [Gene Expression Omnibus repository - GEO] (www.ncbi.nlm.nih.gov/geo)[@GEO].

We used Geometadb[@Geometadb] to perform the search of GEO data sets and RSQLite[@RSQLite] and GEOquerry[@GEOquery] package to perform the queries using a downloaded copy of GEO contents' metadata. These packages were installed, when needed through BiocManager[@BiocManager] functionality. knitr[@knitr] package was later used to report data statistics and edgeR to aid on modeling and normalization of data.

```{r echo = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")

if (!requireNamespace("RSQLite", quietly = TRUE))
  BiocManager::install("RSQLite")

if (!requireNamespace("GEOquery", quietly = TRUE))
  BiocManager::install("GEOquery")

if (!requireNamespace("knitr", quietly = TRUE))
  BiocManager::install("knitr")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
  BiocManager::install("ComplexHeatmap")

if (!requireNamespace("circlize", quietly = TRUE))
  BiocManager::install("circlize")

if(!file.exists('GEOmetadb.sqlite')) 
  GEOmetadb::getSQLiteFile()

library(BiocManager)
library(GEOmetadb)
library(RSQLite)
library(GEOquery)
library(knitr)
library(edgeR)
library(ComplexHeatmap)
library(circlize)
```

```{r eval = FALSE, echo = FALSE}
con <- dbConnect(SQLite(),'GEOmetadb.sqlite')

sql <- paste("SELECT DISTINCT gse.title,gse.gse, gpl.title,",
            " gse.submission_date,",
            " gse.supplementary_file",
            "FROM",
            " gse JOIN gse_gpl ON gse_gpl.gse=gse.gse",
            " JOIN gpl ON gse_gpl.gpl=gpl.gpl",
            "WHERE",
            " gse.submission_date > '2011-02-01' AND",
            " gse.title LIKE '%depressive%' AND",
            " gpl.organism LIKE '%Homo sapiens%' AND",
            " gpl.technology LIKE '%high-throughput sequencing%' ",
            " ORDER BY gse.submission_date DESC",sep=" ")

rs <- dbGetQuery(con,sql)
dim(rs)
dbDisconnect(con)
```

After 4 search iterations we have chosen the following study: [Serotonin-induced hyperactivity in SSRI-resistant major depressive disorder patient-derived neurons] (https://pubmed.ncbi.nlm.nih.gov/30700803/)[@vadodaria2019serotonin] in which the authors tried to elucidate the common clinical challenge of patients suffering from Major Depressive Disorder (MDD) who fail or respond only partially to SSRI therapy (aka non-remitters). We currently have no biomarkers to identify these patients. Moreover, is not known the specific mechanism underpinning SSRI and similar SNRI-induced suicidality.

It also provided the raw mRNA counts for analysis. The controls are represented by expression data in forebrain neuron lines developed from Induced pluripotent stem cells (iPSCs) removed through non-CNS biopsy of healthy patients exposed to SSRIs. The test group involve the same cells from MDD patients which suffered from SSRI treatment resistance and MDD patients which responded to SSRI treatment.

The sample size in this study is 9, with adequate control and 2 testing groups (MDD with an without SSRI-based treatment resistance). 


## 1.2. Downloading the data, cleaning and compute summary statistics

The data for this study is contained in one supplemental .csv file called "GSE125664_Vadodaria_MDDNeurons_RawCounts":

```{r echo=FALSE}
if(!file.exists('SSRI_exp_file')) {
    sfiles = getGEOSuppFiles('GSE125664')
    fnames = rownames(sfiles)
    # Just one file, but we have a .csv here
    SSRI_exp = read.csv(fnames[1],header=TRUE, check.names = FALSE)
    # Naming its gene name column as gname
    names(SSRI_exp)[1] <- "gname"
    saveRDS(SSRI_exp, file = "SSRI_exp_file.rds")
}

if(!exists("SSRI_exp")){
    SSRI_exp <- readRDS(file = "SSRI_exp_file.rds")
}

# Capturing its description, inspired on lecture 4
SSRI_gse <- getGEO("GSE125664",GSEMatrix=FALSE)
SSRI_gpl <- names(GPLList(SSRI_gse))[1]
SSRI_gpl_info <- Meta(getGEO(SSRI_gpl))
```

__Platform title:__ `r SSRI_gpl_info$title`

__Submission date:__ `r SSRI_gpl_info$submission_date`

__Last update date:__ `r SSRI_gpl_info$last_update_date`

__Organisms:__  `r SSRI_gpl_info$organism`

Snapshot of data first 10 mapped genes at this stage:

```{r fig.width= 3, fig.height= 3}
kable(SSRI_exp[1:10,1:10], format = "html")
```

```{r}
dim(SSRI_exp)
```

And we have 22351 genes with measures for 3 controls, 3 non-resistant-to-SSRI MDD patients and 3 resistant-to-SSRI MDD patients.

Counts for each gene show each gene with the frequency of exactly one, except for some miscoded names.

```{r eval = FALSE, echo = FALSE}
gene_counts <- sort(table(SSRI_exp$gname), decreasing = TRUE)
```

28 genes contained miscoded names (they expressed dates instead of valid identifiers) and we will have to remove them. 

```{r echo = FALSE}
SSRI_exp_clean <- SSRI_exp[ !(SSRI_exp$gname %in% c("1-Mar", "1-Mar", "2-Mar", "2-Mar", "1-Dec", "1-Sep", "10-Mar", "10-Sep", "11-Mar", "11-Sep", "12-Sep", "14-Sep", "15-Sep", "2-Sep", "3-Mar", "3-Sep", "4-Mar", "4-Sep", "5-Mar", "5-Sep", "6-Mar", "6-Sep", "7-Mar", "7-Sep", "8-Mar", "8-Sep", "9-Mar", "9-Sep")), ]
```

Each mRNA reading was already mapped to unique genes expressed as HUGO Gene Nomenclature Committee symbols) so we proceeded to calculate counts per million reads mapped (cmp). All the counts with less than 9 counts where removed and that removed 9350 gene observations. There after that, and after the cleaning the data, there were no duplicates and all the symbols where unique symbols that complied with HUGO Gene Nomenclature Committee. After all removals and manipulations, the dataset coverage was 0.59.

```{r echo = FALSE}
cpms = cpm(SSRI_exp_clean[,2:10])
rownames(cpms) <- SSRI_exp_clean[,1]
# Our study contains 9 samples
keep = rowSums(cpms >1) >=9
SSRI_exp_filtered = SSRI_exp_clean[keep,]
```

We have kept 12973 gene counts, meaning we had to filter out `r 1-(dim(SSRI_exp_filtered)[1]/dim(SSRI_exp_clean)[1])` of our gene-mapped RNA seq observations

## 1.3. Normalization

We used TMM (Trimmed Mean of M-values) normalization. For most samples the distribution curve pre and post-normalization is very similar, however there is a noticiable change towards the mean (a bit higher than 5) for the second control (H_neuron_2) and, interestingly, for 2 of the 3 nonremitter patients.

```{R fig.width=5, fig.height=5}

data2plot <- log2(cpm(SSRI_exp_filtered[,2:10]))

SSRI_matrix <- as.matrix(SSRI_exp_filtered[,-1])

rownames(SSRI_matrix) <- SSRI_exp_filtered$gname

# inspired by Yi Fei Huang approach in terms of merging samples into groups

categories <- c("healthy", "healthy", "healthy",
             "nonremitter", "nonremitter", "nonremitter",
             "remitter", "remitter", "remitter")

# inspired by lecture 4 slides 51 and 52

d = DGEList(counts=SSRI_matrix, group=categories)

# Calculation of normalization factors

d = calcNormFactors(d)

normalized_counts <- cpm(d)
    
```



# 2. Differential Gene Expression analysis

## 2.1 Model design and Differential Expression Significance Calculation with correction

In this section, we will used the normalized data produced above to test for statistically significant expression differences for each of the mapped genes, between our 3 groups (we will call it status) of interest in our designed model: healthy patients exposed to SSRI (control), non-remitters MDD exposed to SSRI (test 2) and remitter MDD exposed to SSRI (test 2). 

An expansion of this model includes 2 groups (we will call this attribute "affected"), healthy (ND) and affected by disease (D), although this expansion wouldn't add granularity to the assessment of different subtypes of MDD (refractive or not to treatment), it could help in increasing specificity to the bayes differential expression significance calculation.

```{r}
# First we create our model

samples <- data.frame(status = c("H", "H", "H", "NR", "NR","NR", "R", "R", "R"),
                      affected = c("ND", "ND", "ND", "D", "D","D", "D", "D", "D"))

rownames(samples) <- colnames(SSRI_exp[-1])
# model on affected Vs healthy
model_design <- model.matrix(~samples$affected)
# model on status of remission to SSRI (remitters Vs non-remitters)
model_design_2 <- model.matrix(~samples$status)

# compound model taking into account both conditions will not yield independent coefficients:
# model_design_compound <- model.matrix(~samples$affected + samples$status)
  
```

Now we will fit our normalized data to each of the built models.

### 2.1.1 Using Limma
```{r}
# As inspired by lecture 6, using limma, doesn't add result in significantly different
# restults compared with chosen approach (see following code chunk):
expressionMatrix <- as.matrix(normalized_counts)
rownames(expressionMatrix) <- SSRI_exp_filtered$gname
colnames(expressionMatrix) <- colnames(SSRI_exp_filtered)[2:10]
minimalSet <- ExpressionSet(assayData=expressionMatrix)
# Fitting for Affected Vs Controls
fit <- lmFit(minimalSet, model_design)
# Fitting for type of SSRI response (remitters and non-remitters)
fit_resp <- lmFit(minimalSet, model_design_2)
```
Using empirical Bayes method:
```{r}
fit_2 <- eBayes(fit_resp,trend=TRUE)
fit_resp_2 <- eBayes(fit_resp,trend=TRUE)
```
Now we can calculate the differential expression and apply the correction for multiple
hypothesis test through Benjamni-Hochberg method (less than 1 minute processing time).
```{r}
#calculate differential expression, inspired by lecture 6
topfit <- topTable(fit_2, coef=ncol(model_design), adjust.method = "BH",
number = nrow(expressionMatrix))

topfit_resp <- topTable(fit_resp_2, coef=ncol(model_design_2), adjust.method = "BH",
number = nrow(expressionMatrix))

gene_names <- data.frame(SSRI_exp[,1])
colnames(gene_names) <- c("hgnc_symbol")
output_hits <- merge(gene_names, topfit, by.y=0, by.x=1, all.y=TRUE)
output_hits_2 <- merge(gene_names, topfit_resp, by.y=0, by.x=1, all.y=TRUE)

#sort by p-value
output_hits <- output_hits[order(output_hits$P.Value),]
output_hits_2 <- output_hits_2[order(output_hits_2$P.Value),]
```
And the results for each of the models, with and without Benjamni-Hochberg corrections
for multiple hypothesis tests are as following.

For Healthy Vs MDD patients:

```{r}
kable(output_hits[1:10,],type="html",row.names = FALSE)
```
For Remitter status:
```{r}
kable(output_hits_2[1:10,],type="html",row.names = FALSE)
```
Applying a threshold of p-value < 0.05 and computing how many genes passed the test:
```{r}
length(which(output_hits$P.Value < 0.05))
length(which(output_hits_2$P.Value < 0.05))
```
* 2612 genes passed the test of significance for change in transcription factor for 
  the healthy VS MDD model
* 983 genes passed the test of significance for change in transcription factor for 
  the "type-of-response-to-SSRI model


Now computing how many mapped genes passed the p-value < 0.05 with Benjamni-Hochberg correction for multiple testing
```{r}
length(which(output_hits$adj.P.Val < 0.05))
length(which(output_hits_2$adj.P.Val < 0.05))

```
* No gene passed the test of significance for change in transcription factor for 
  the healthy VS MDD model with correction
* No gene passed the test of significance for change in transcription factor for 
  the "type-of-response-to-SSRI model with correction


### 2.1.2 EdgeR Simple approach using one model
In this approach we will use the model that differentiates between MDD patients that were
non-responsive to SSRI (non-remitter or NR) from responsive patients (remitter or R), since
this is main approach took by the paper's authors in its published version's transcriptome analysis (see [@vadodaria2019serotonin], page 801, specially fig 4.b)
```{r}
d = DGEList(counts=normalized_counts, group=samples$statusNR)

# Dispersions
d <- estimateDisp(d, model_design_2)

# Computing normalization factors
d <- calcNormFactors(d)

# fitting models
fit <- glmQLFit(d, model_design_2)

# Computing differential expression
qlf.pos_vs_neg <- glmQLFTest(fit, coef='samples$statusNR')
```

```{r}
# Results
qlf_output_hits <- topTags(qlf.pos_vs_neg, sort.by = "PValue", n = nrow(SSRI_exp))
# Passing p-value < 0.05?
length(which(qlf_output_hits$table$PValue < 0.05))
# Passing correction
length(which(qlf_output_hits$table$FDR < 0.05))
## [1] 0
```
* 2536 genes passed the test of significance for change in transcription
* No gene passed the test of significance for change in transcription

### 2.1.3 Using paper's modeling

Since the previous 2 modeling failed to yield significant results post-correction, we decided
to use the model applied by the source study [@vadodaria2019serotonin], page 801. Thus, we design 2 distinct but symmetrical models that will assess remitters MDD patient's cells VS Healthy control cells and remitters MDD patient's cells VS non-remitters MDD patient's cells separately.

samples <- data.frame(status = c("H", "H", "H", "NR", "NR","NR", "R", "R", "R"),
                      affected = c("ND", "ND", "ND", "D", "D","D", "D", "D", "D"))

```{r}

# Creating model for non-remitters VS control assessment
samples_NRxH <- data.frame(status = c("H", "H", "H", "NR", "NR","NR"))
rownames(samples_NRxH) <- colnames(SSRI_exp)[2:7]
model_design_NRxH <- model.matrix(~samples_NRxH$status)

# Creating model for non-remitters VS remitters assessment
samples_NRxR <- data.frame(status = c("NR", "NR","NR", "R", "R", "R"))
rownames(samples_NRxR) <- colnames(SSRI_exp)[5:10]
model_design_NRxR <- model.matrix(~samples_NRxR$status)
```

Now we apply the each model separately to the appropriate data set:

```{r}
d_NRxH = DGEList(counts=normalized_counts[,1:6], group=samples_NRxH$statusNR)
d_NRxR = DGEList(counts=normalized_counts[,4:9], group=samples_NRxR$statusNR)


# Dispersions
d_NRxH <- estimateDisp(d_NRxH, model_design_NRxH)
d_NRxR <- estimateDisp(d_NRxR, model_design_NRxR)


# Computing normalization factors
d_NRxH <- calcNormFactors(d_NRxH)
d_NRxR <- calcNormFactors(d_NRxR)


# fitting models
fit_NRxH <- glmQLFit(d_NRxH, model_design_NRxH)
fit_NRxR <- glmQLFit(d_NRxR, model_design_NRxR)


# Computing differential expression
qlf.pos_vs_neg_NRxH <- glmQLFTest(fit_NRxH, coef='samples_NRxH$statusNR')
qlf.pos_vs_neg_NRxR <- glmQLFTest(fit_NRxR, coef='samples_NRxR$statusR')
```


```{r}
# Results, now inspired by lecture 7
qlf_output_hits_RNxH <- topTags(qlf.pos_vs_neg_NRxH, sort.by = "PValue", n = nrow(SSRI_exp))
qlf_output_hits_RNxR <- topTags(qlf.pos_vs_neg_NRxR, sort.by = "PValue", n = nrow(SSRI_exp))

# Here we check how many mapped genes passed the 0.05 significance threshold with and without correction
# for the comparison remitter Vs control
length(which(qlf_output_hits_RNxH$table$PValue < 0.05))
length(which(qlf_output_hits_RNxH$table$FDR < 0.05))

# Here we check how many mapped genes passed the 0.05 significance threshold with and without correction
# for the comparison non-remitters Vs remitters
length(which(qlf_output_hits_RNxR$table$PValue < 0.05))
length(which(qlf_output_hits_RNxR$table$FDR < 0.05))
```

Although the results indicate the need for enrichment set analysis (which is one of the purpose of this assignment,
as follows), this results are now more comparable with the source literature[@vadodaria2019], in which, at least
for the set of genes presented there, the differential expression between remitters and controls is larger than
the one between remitters and non-remitters. 

Another thresholding, now with p=0.09 is now performed, despite not being considered as a strong evidence of up or downregulation. We decided to perform this aditional less stringent comparision in order to compare with the original paper's finding of HTR set (specifically HTR2A) are up-regulated in non-remitters compared to remmiters.

```{r}
length(which(qlf_output_hits_RNxR$table$PValue < 0.09))
```
And it increases drastically the amount of hits, compatible with what the paper presented, although we won't proceed further with the 0.09 threshold in order to keep relevance to any eventual finding.

## 2.2 General and highlighted MA plot 

Our top results through the last model were as follows, for the remitters VS control comparison:

```{r}
kable(topTags(qlf.pos_vs_neg_NRxH), type="html")
```

And for the non-remitters VS remitters comparision:

```{r}
kable(topTags(qlf.pos_vs_neg_NRxR), type="html")
```


In order to have a more general view of the differences in expression among non-remitters and controls and among remitters and non-remitters we decided to build an MA plot for each of these assessments. For that, we will use plotMA from the limma package[@limma]. We will also use the <status> argument at the function (see documentation), to include the controls.

Comparing first pair of samples (remmitter Vs non-remmiters and its control)

```{r fig.width= 4 fig.height=4}
plotMA(log2(SSRI_exp[,c(5,8)]), ylab="M - ratio log expression", main="MA plot R VS NR 1st pair", status = SSRI_exp[,c(2)])
```
Second pair:
```{r fig.width= 4 fig.height=4}
plotMA(log2(SSRI_exp[,c(6,9)]), ylab="M - ratio log expression", main="MA plot R VS NR 2nd pair", status = SSRI_exp[,c(3)])
```

Third pair:
```{r fig.width= 4 fig.height=4}
plotMA(log2(SSRI_exp[,c(7,10)]), ylab="M - ratio log expression", main="MA plot R VS NR 3rd pair", status = SSRI_exp[,c(3)])
```

No useful insight is prompt when such large number of gene mapping is available so we decided to highlight only the group of genes related to the paper's biological hypothesis, concerning the following HTR family genes:



## Heat map

In order to vizualize our mostly deferentially expressed genes we used a heat map as provided by the ComplexHeatmap package[@ComplexHeatmap)], also using circlize package [@circlize]. However the processing time was long for the whole dataset using those tools.

```{r eval = FALSE}
# As inspired in Lecture 6

# Creating matrix
heatmap_matrix <- normalized_counts[,2:ncol(normalized_counts)]

rownames(heatmap_matrix) <- rownames(normalized_counts)

colnames(heatmap_matrix) <- colnames(normalized_counts[,2:ncol(normalized_counts)])

heatmap_matrix <- t(scale(t(heatmap_matrix)))

# Setting colours
if(min(heatmap_matrix) == 0){
  heatmap_col = colorRamp2(c( 0, max(heatmap_matrix)),
c( "white", "red"))
} else {
heatmap_col = colorRamp2(c(min(heatmap_matrix), 0,
max(heatmap_matrix)), c("blue", "white", "red"))
}

# Creating heat map
current_heatmap <- Heatmap(as.matrix(heatmap_matrix),
show_row_dend = TRUE,show_column_dend = TRUE,
col=heatmap_col,show_column_names = TRUE,
show_row_names = FALSE,show_heatmap_legend = TRUE)
current_heatmap
```

So we decided to use the limma package[@limma] tools instead, using a threshold of 0.01 for significance for a cleaner view.

First for non-remitters VS controls:

```{r}
# As inspired by lecture 6

heatmap_matrix <- normalized_counts[,1:6]

rownames(heatmap_matrix) <- rownames(normalized_counts)

colnames(heatmap_matrix) <- colnames(normalized_counts[,1:6])

heatmap_matrix <- t(scale(t(heatmap_matrix)))

top_hits <- output_hits_2$hgnc_symbol[output_hits_2$P.Value<0.01]

heatmap_matrix_tophits <- t(scale(t(heatmap_matrix[which(rownames(heatmap_matrix)
                          %in% top_hits),])))

# UNCOMMENT THIS TO HIDE CLUSTERING
# This is to specifically avoid clustering by condition
# heatmap_matrix_tophits<- heatmap_matrix_tophits[,
# c(grep(colnames(heatmap_matrix_tophits),pattern = "1"),
# grep(colnames(heatmap_matrix_tophits),pattern = "2"),
# grep(colnames(heatmap_matrix_tophits),pattern = "3"))]
  
if(min(heatmap_matrix_tophits) == 0){
  heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
  c( "white", "red"))
} else {
  heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0, 
                          max(heatmap_matrix_tophits)), 
                          c("blue", "white", "red"))
}
current_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
cluster_rows = TRUE, show_row_dend = TRUE,
cluster_columns = FALSE,show_column_dend = FALSE,
col=heatmap_col,show_column_names = TRUE,
show_row_names = FALSE,show_heatmap_legend = TRUE)

current_heatmap

```
Legend: The conditions here (status of non-remmiter or control) are intentionally clusted toguether due to the order of the matrix (cluster argument still set to FALSE) used on the heat map (uncomment unclustering code to see it otherwise). One can notice that the bottom half of the over expressed sets on the non-remmiter band is more active at remitters neurons 2 and 3 but still overexpressed across all test samples.


Now for non-remitters VS remitters:

```{r}
# As inspired by lecture 6

heatmap_matrix <- normalized_counts[,4:9]

rownames(heatmap_matrix) <- rownames(normalized_counts)

colnames(heatmap_matrix) <- colnames(normalized_counts[,4:9])

heatmap_matrix <- t(scale(t(heatmap_matrix)))

top_hits <- output_hits_2$hgnc_symbol[output_hits_2$P.Value<0.01]

heatmap_matrix_tophits <- t(scale(t(heatmap_matrix[which(rownames(heatmap_matrix)
                          %in% top_hits),])))

# UNCOMMENT THIS TO HIDE CLUSTERING
# This is to specifically avoid clustering by condition
# heatmap_matrix_tophits<- heatmap_matrix_tophits[,
# c(grep(colnames(heatmap_matrix_tophits),pattern = "1"),
# grep(colnames(heatmap_matrix_tophits),pattern = "2"),
# grep(colnames(heatmap_matrix_tophits),pattern = "3"))]
  
if(min(heatmap_matrix_tophits) == 0){
  heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
  c( "white", "red"))
} else {
  heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0, 
                          max(heatmap_matrix_tophits)), 
                          c("blue", "white", "red"))
}
current_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
cluster_rows = TRUE, show_row_dend = TRUE,
cluster_columns = FALSE,show_column_dend = FALSE,
col=heatmap_col,show_column_names = TRUE,
show_row_names = FALSE,show_heatmap_legend = TRUE)

current_heatmap

```

The conditions here (status of remitter or non-remitter) are intentionally clusted toguether due to the order of the matrix used on the heat map (cluster argument still set to FALSE). One can notice that except for a mid-range band accross all NR's and except for the lower half of NR_neuron_1, the non-remmiter are  active than the correspondent band on the remitter neurons.

# 3. Thresholded over-representation analysis

## 3.1 Quantifying significantly up and down-regulated genes

First comparing non-remitters and controls:
```{r}
## up regulated genes:
length(which(qlf_output_hits_RNxH$table$PValue < 0.05
& qlf_output_hits_RNxH$table$logFC > 0))
## down regulated genes:
length(which(qlf_output_hits_RNxH$table$PValue < 0.05
& qlf_output_hits_RNxH$table$logFC < 0))
```
We have 1345 genes significantly up-regulated in non-remmiters comparing to controls, and 961 down-regulated.

Now comparing non-remitters and remitters:
```{r}
## up regulated genes:
length(which(qlf_output_hits_RNxR$table$PValue < 0.05
& qlf_output_hits_RNxR$table$logFC > 0))
## down regulated genes:
length(which(qlf_output_hits_RNxR$table$PValue < 0.05
& qlf_output_hits_RNxR$table$logFC < 0))
```
We have 379 genes significantly up-regulated in non-remmiters comparing to controls, and 353 down-regulated, which is compatible with finding on paper, where the delta of expression among MDD patients is decreased when compared to remitters and controls.

## 3.2 Creating thresholded list

Here we will create a thresholded list, initially with the remitters dataset compared
to the non remitters dataset, in which our rank will consist in the following product:
-log(p-value) * logFC. We then create 3 files to contain, separately:

1) all significantly differentially expressed genes together 
2) the up-regulated significantly differentially expressed genes
3) the down-regulated significantly differentially expressed genes

```{r}
# Inspired by lecture 7

# Attached name to gene results matrix
qlf_output_hits_withgn_RNxR <- merge(SSRI_exp[,1], qlf_output_hits_RNxR, by.x=1, by.y=0)

colnames(qlf_output_hits_withgn_RNxR) <- c("gname", colnames(qlf_output_hits_withgn_RNxR)[-1])

# Calculating our rank as -log(p-value) * logFC
qlf_output_hits_withgn_RNxR[,"rank"] <- -log(qlf_output_hits_withgn_RNxR$PValue,base=10) * 
                                         sign(qlf_output_hits_withgn_RNxR$logFC)

# Ordering per rank
qlf_output_hits_withgn_RNxR <- qlf_output_hits_withgn_RNxR[order(qlf_output_hits_withgn_RNxR$rank),]

# Separating up-regulated genes with threshold of 0.05
# Note: the previous calculation modeled for FC R/NR
# here, we are interested in FC NR/R
upregulated_genes_RNxR <- qlf_output_hits_withgn_RNxR$gname[which(qlf_output_hits_withgn_RNxR$PValue < 0.05
& qlf_output_hits_withgn_RNxR$logFC < 0)]

# Separating down-regulated genes with threshold of 0.05
downregulated_genes_RNxR <- qlf_output_hits_withgn_RNxR$gname[which(qlf_output_hits_withgn_RNxR$PValue 
                                                    < 0.05 & qlf_output_hits_withgn_RNxR$logFC > 0)]
```

Now doing the same to compare non-remitters to remitters:
```{r}
# Inspired by lecture 7

# Attached name to gene results matrix
qlf_output_hits_withgn_RNxH <- merge(SSRI_exp[,1], qlf_output_hits_RNxH, by.x=1, by.y=0)

colnames(qlf_output_hits_withgn_RNxH) <- c("gname", colnames(qlf_output_hits_withgn_RNxH)[-1])

# Calculating our rank as -log(p-value) * logFC
qlf_output_hits_withgn_RNxH[,"rank"] <- -log(qlf_output_hits_withgn_RNxH$PValue,base=10) * 
                                         sign(qlf_output_hits_withgn_RNxH$logFC)

# Ordering per rank
qlf_output_hits_withgn_RNxH <- qlf_output_hits_withgn_RNxH[order(qlf_output_hits_withgn_RNxH$rank),]

# Separating up-regulated genes with threshold of 0.05
upregulated_genes_RNxH <- qlf_output_hits_withgn_RNxH$gname[which(qlf_output_hits_withgn_RNxH$PValue < 0.05
& qlf_output_hits_withgn_RNxH$logFC > 0)]

# Separating down-regulated genes with threshold of 0.05
downregulated_genes_RNxH <- qlf_output_hits_withgn_RNxH$gname[which(qlf_output_hits_withgn_RNxH$PValue 
                                                    < 0.05 & qlf_output_hits_withgn_RNxH$logFC < 0)]
```


Save tables for later use if docker port connection is properly configured:
```{r}
dir.create('output', showWarnings=FALSE)

# Saving up-regulated genes NRxR
write.table(x=upregulated_genes_RNxR, file=file.path("./","RNxR_upregulated_genes.txt"),sep="\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)

# Saving down-regulated genes NRxR
write.table(x=downregulated_genes_RNxR, file=file.path("./","RNxR_downregulated_genes.txt"),sep="\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)

# Saving complete list NRxR
write.table(x=data.frame(genename=qlf_output_hits_withgn_RNxR$gname,
                         F_stat=qlf_output_hits_withgn_RNxR$rank
                         ),
                         file=file.path("./","RNxR_ranked_genelist.txt"),
                         sep = "\t",
                         row.names = FALSE,col.names = FALSE,quote = FALSE)

# Saving up-regulated genes RNxH
write.table(x=upregulated_genes_RNxH, file=file.path("./","RNxH_upregulated_genes.txt"),sep="\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)

# Saving down-regulated genes RNxH
write.table(x=downregulated_genes_RNxH, file=file.path("./","RNxH_downregulated_genes.txt"),sep="\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)

# Saving complete list RNxH
write.table(x=data.frame(genename=qlf_output_hits_withgn_RNxH$gname,
                         F_stat=qlf_output_hits_withgn_RNxH$rank
                         ),
                         file=file.path("./","RNxH_ranked_genelist.txt"),
                         sep = "\t",
                         row.names = FALSE,col.names = FALSE,quote = FALSE)
```

## 3.3 Gene List Function Enrichment (over-representation) Analysis

We will be using G:Profiler web server [@GProfiler]to perform the above gene list enrichment analysis (version e102_eg49_p15_7a9b4d6).

Thus we will be using the following annotation database in the enrichment:

- Through GO biological process annotation release 2021-02-01
  (with Ensembl version 102)
- Reactome version 75 (December 2020)
- Wikipathways (December 2020 release)

All electronic GO annotations were discarded and the annotation term size was restricted to 200 genes.

The tool uses Fisher Exact test. We used 0.05 threshold for significance. In our list of up and down regulated genes we could filter for only 4 members of HTR gene family, since we had a strict standard to cut out any mapped genes with less than 9 reads. Thus we have now decided to use Benjamni-Hochberg correction for multiple hypothesis testing due to its less stringent profile compared to Bonferoni correction.

# References


